aten/src/ATen/native/cuda/group_norm_kernel.cu

```diff
@@ -149,6 +183,199 @@ __global__ void ComputeGradOutputCoeffientCUDAKernel(
+    sum1 = cuda_utils::BlockReduceSum<T_ACC>(sum1, ds_shared);
+    sum2 = cuda_utils::BlockReduceSum<T_ACC>(sum2, db_shared);
+  }
+  if (threadIdx.x == 0) {
+    const T_ACC s = T_ACC(1) / static_cast<T_ACC>(D);
+    const T_ACC x = (sum2 * static_cast<T_ACC>(mean[ng]) - sum1) *
+        static_cast<T_ACC>(rstd[ng]) * static_cast<T_ACC>(rstd[ng]) *
+        static_cast<T_ACC>(rstd[ng]) * s;
+    c2[ng] = x;
+    c3[ng] = -x * static_cast<T_ACC>(mean[ng]) -
+        sum2 * static_cast<T_ACC>(rstd[ng]) * s;
+  }
+}
+
+template <typename T>
+__global__ void GroupNorm1dBackwardCUDAKernel(
+    int64_t N,
+    int64_t C,
+    int64_t group,
+    const T* dY,
+    const T* X,
+    const T* rstd,
+    const T* gamma,
+    const acc_type<T, true>* c2,
+    const acc_type<T, true>* c3,
+    T* dX) {
+  using T_ACC = acc_type<T, true>;
+  const int64_t index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index < N * C) {
+    const int64_t c = index % C;
+    const int64_t ng = index / (C / group);
+    const T_ACC c1 = static_cast<T_ACC>(rstd[ng]) *
+        (gamma == nullptr ? T_ACC(1) : static_cast<T_ACC>(gamma[c]));
+    dX[index] = c1 * static_cast<T_ACC>(dY[index]) +
+        c2[ng] * static_cast<T_ACC>(X[index]) + c3[ng];
+  }
+}
+
+template <typename T>
+__global__ void GammaBeta1dBackwardCUDAKernel1(
+    int64_t N,
+    int64_t C,
+    int64_t group,
+    const T* dY,
+    const T* X,
+    const T* mean,
+    const T* rstd,
+    T* dgamma,
+    T* dbeta) {
+  using T_ACC = acc_type<T, true>;
+  const int64_t c = blockIdx.x * blockDim.x + threadIdx.x;
+  if (c < C) {
+    const int64_t G = group;
+    const int64_t D = C / G;
+    T_ACC sum1 = 0;
+    T_ACC sum2 = 0;
+    for (int64_t n = 0; n < N; ++n) {
+      const int64_t nc = n * C + c;
+      const int64_t ng = n * G + c / D;
+      const T_ACC dy_acc = static_cast<T_ACC>(dY[nc]);
+      const T_ACC x_acc = static_cast<T_ACC>(X[nc]);
+      sum1 += (dgamma == nullptr)
+          ? T_ACC(0)
+          : ((dy_acc * x_acc - dy_acc * static_cast<T_ACC>(mean[ng])) *
+             static_cast<T_ACC>(rstd[ng]));
+      sum2 += (dbeta == nullptr) ? T_ACC(0) : dy_acc;
+    }
+    if (dgamma != nullptr) {
+      dgamma[c] = sum1;
+    }
+    if (dbeta != nullptr) {
+      dbeta[c] = sum2;
+    }
+  }
+}
+
+template <typename T>
+__global__ void GammaBeta1dBackwardCUDAKernel2(
+    int64_t N,
+    int64_t C,
+    int64_t group,
+    const T* dY,
+    const T* X,
+    const T* mean,
+    const T* rstd,
+    T* dgamma,
+    T* dbeta) {
+  using T_ACC = acc_type<T, true>;
+  __shared__ T_ACC g_shared[kReduceTileSize][kReduceTileSize + 1];
+  __shared__ T_ACC b_shared[kReduceTileSize][kReduceTileSize + 1];
+  const int64_t c = blockIdx.x * blockDim.x + threadIdx.x;
+  T_ACC dg_sum1 = 0;
+  T_ACC dg_sum2 = 0;
+  T_ACC db_sum1 = 0;
+  T_ACC db_sum2 = 0;
+  if (c < C) {
+    const int64_t G = group;
+    const int64_t D = C / G;
+    for (int64_t n = threadIdx.y; n < N; n += blockDim.y * 2) {
```

Comment:
can you please leave brief comments on what this kernel is doing? Like, you unroll by a factor of 2, half of shared memory is to keep the first part of unrolled loads, half to keep the second etc (would it make sense to give another dimension to shared mem?)

Reply:
Done.